url11fb0973cf_raichuman9_Swimmer526_tgz-pJ8-fr_inria_optimization_cmaes_examples_CMAExample1J8/src/org/jbox2d/dynamics/joints/RevoluteJoint.java

```diff
@@ -0,0 +1,582 @@
+		Mat22.mulToOut(b1.getTransform().R, r1, r1);
+		Mat22.mulToOut(b2.getTransform().R, r2, r2);
+		
+		// J = [-I -r1_skew I r2_skew]
+		// [ 0 -1 0 1]
+		// r_skew = [-ry; rx]
+		
+		// Matlab
+		// K = [ m1+r1y^2*i1+m2+r2y^2*i2, -r1y*i1*r1x-r2y*i2*r2x, -r1y*i1-r2y*i2]
+		// [ -r1y*i1*r1x-r2y*i2*r2x, m1+r1x^2*i1+m2+r2x^2*i2, r1x*i1+r2x*i2]
+		// [ -r1y*i1-r2y*i2, r1x*i1+r2x*i2, i1+i2]
+		
+		// K = [(1/m1 + 1/m2) * eye(2) - skew(r1) * invI1 * skew(r1) - skew(r2) * invI2 *
+		// skew(r2)]
+		// = [1/m1+1/m2 0 ] + invI1 * [r1.y*r1.y -r1.x*r1.y] + invI2 * [r1.y*r1.y
+		// -r1.x*r1.y]
+		// [ 0 1/m1+1/m2] [-r1.x*r1.y r1.x*r1.x] [-r1.x*r1.y r1.x*r1.x]
+		
+		float m1 = b1.m_invMass, m2 = b2.m_invMass;
+		float i1 = b1.m_invI, i2 = b2.m_invI;
+		
+		m_mass.col1.x = m1 + m2 + r1.y * r1.y * i1 + r2.y * r2.y * i2;
+		m_mass.col2.x = -r1.y * r1.x * i1 - r2.y * r2.x * i2;
+		m_mass.col3.x = -r1.y * i1 - r2.y * i2;
+		m_mass.col1.y = m_mass.col2.x;
+		m_mass.col2.y = m1 + m2 + r1.x * r1.x * i1 + r2.x * r2.x * i2;
+		m_mass.col3.y = r1.x * i1 + r2.x * i2;
+		m_mass.col1.z = m_mass.col3.x;
+		m_mass.col2.z = m_mass.col3.y;
+		m_mass.col3.z = i1 + i2;
+		
+		m_motorMass = i1 + i2;
+		if (m_motorMass > 0.0f) {
+			m_motorMass = 1.0f / m_motorMass;
+		}
+		
+		if (m_enableMotor == false) {
+			m_motorImpulse = 0.0f;
+		}
+		
+//		System.out.printf("joint angle: %f, %f, and %f\n", b2.m_sweep.a, b1.m_sweep.a, m_referenceAngle);
+		
+		if (m_enableLimit) {
+			float jointAngle = b2.m_sweep.a - b1.m_sweep.a - m_referenceAngle;
+			if (MathUtils.abs(m_upperAngle - m_lowerAngle) < 2.0f * Settings.angularSlop) {
+				m_limitState = LimitState.EQUAL;
+			}
+			else if (jointAngle <= m_lowerAngle) {
+				if (m_limitState != LimitState.AT_LOWER) {
+					m_impulse.z = 0.0f;
+				}
+				m_limitState = LimitState.AT_LOWER;
+			}
+			else if (jointAngle >= m_upperAngle) {
+				if (m_limitState != LimitState.AT_UPPER) {
+					m_impulse.z = 0.0f;
+				}
+				m_limitState = LimitState.AT_UPPER;
+			}
+			else {
+				m_limitState = LimitState.INACTIVE;
+				m_impulse.z = 0.0f;
+			}
+		}
+		else {
+			m_limitState = LimitState.INACTIVE;
+		}
+//		System.out.printf("limit state: %s\n", m_limitState.toString());
+		
+		if (step.warmStarting) {
+			// Scale impulses to support a variable time step.
+			m_impulse.mulLocal(step.dtRatio);
+			m_motorImpulse *= step.dtRatio;
+			
+			Vec2 temp = pool.popVec2();
+			Vec2 P = pool.popVec2();
+			P.set(m_impulse.x, m_impulse.y);
+			
+			temp.set(P).mulLocal(m1);
+			b1.m_linearVelocity.subLocal(temp);
+			b1.m_angularVelocity -= i1 * (Vec2.cross(r1, P) + m_motorImpulse + m_impulse.z);
+			
+			temp.set(P).mulLocal(m2);
+			b2.m_linearVelocity.addLocal(temp);
+			b2.m_angularVelocity += i2 * (Vec2.cross(r2, P) + m_motorImpulse + m_impulse.z);
+			
+			pool.pushVec2(2);
+		}
+		else {
+			m_impulse.setZero();
+			m_motorImpulse = 0.0f;
+		}
+		pool.pushVec2(2);
+	}
+	
+	@Override
+	public void solveVelocityConstraints(final TimeStep step) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 89 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this method or break it into multiple methods.